{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import csv\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "#from Bio.Alphabet import generic_dna\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqIO.QualityIO import FastqGeneralIterator\n",
    "\n",
    "from itertools import chain\n",
    "from typing import TextIO\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_f = '../data/d14_plasmid_library_ratio_targeting_library.csv'\n",
    "df = pd.read_csv(input_f,header=None)\n",
    "df.columns =['guide','gene', 'gene_id', 'pos', 'raw ratio']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_f = '../data/integrated_guide_feature_filtered_f24_mismatch3_all_features.csv'\n",
    "df = pd.read_csv(input_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## guide free energy, Linearfold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make linearfold input\n",
    "standard_prefix = \"CAAGTAAACCCCTACCAACTGGTCGGGGTTTGAAAC\" # DR\n",
    "\n",
    "num_examples = len(df['guide'].values)  \n",
    "with open(\"../data/Linearfold_input/essential_guides_input.fasta\", 'w') as file:\n",
    "    for i in range(num_examples):\n",
    "        file.write(\">guide #\" + str(i + 1) + \"\\n\")\n",
    "        whole_guide = standard_prefix + df['guide'].values[i]\n",
    "        file.write(whole_guide.upper().replace('T','U') + \"\\n\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "! cat ../data/Linearfold_input/essential_guides_input.fasta | ../Linearfold/./linearfold  > ../data/Linearfold_output/essential_guides_output.txt "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_guide_linearfold_fasta_into_dict_contrafold(fname):\n",
    "    fasta_file = open(fname)\n",
    "    seq_dict = {}\n",
    "    score_dict = {}\n",
    "\n",
    "    def parse_one_example(fasta: TextIO):\n",
    "        descr_line = fasta.readline()\n",
    "        if not descr_line:\n",
    "            return None, None, None\n",
    "        guide_seq = fasta.readline().strip()[36:]\n",
    "        linearfold_and_result = fasta.readline()\n",
    "        match = re.match('([\\\\.|\\\\(|\\\\)]+) \\((\\-?[0-9]*\\.[0-9]+)\\)', linearfold_and_result)\n",
    "        linseq, score = match.groups()\n",
    "        score = float(score)\n",
    "\n",
    "        assert '>' in descr_line\n",
    "\n",
    "        return guide_seq, linseq, score\n",
    "\n",
    "    while True:\n",
    "        key, seq, score = parse_one_example(fasta_file)\n",
    "        if key is None:\n",
    "            break\n",
    "        seq_dict[key] = seq\n",
    "        score_dict[key] = score\n",
    "\n",
    "    fasta_file.close()\n",
    "\n",
    "    return seq_dict, score_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_guide_linearfold_fasta_into_dict_contrafold_rna(fname):\n",
    "    fasta_file = open(fname)\n",
    "    seq_dict = {}\n",
    "    score_dict = {}\n",
    "\n",
    "    def parse_one_example(fasta: TextIO):\n",
    "        descr_line = fasta.readline()\n",
    "        if not descr_line:\n",
    "            return None, None, None\n",
    "        guide_seq = fasta.readline().strip()[36:]\n",
    "        linearfold_and_result = fasta.readline()\n",
    "        match = re.match('([\\\\.|\\\\(|\\\\)]+) \\((\\-?[0-9]*\\.[0-9]+)\\)', linearfold_and_result)\n",
    "        linseq, score = match.groups()\n",
    "        score = float(score)\n",
    "\n",
    "        assert '>' in descr_line\n",
    "\n",
    "        return guide_seq.replace('U','T'), linseq, score\n",
    "\n",
    "    while True:\n",
    "        key, seq, score = parse_one_example(fasta_file)\n",
    "        if key is None:\n",
    "            break\n",
    "        seq_dict[key] = seq\n",
    "        score_dict[key] = score\n",
    "\n",
    "    fasta_file.close()\n",
    "\n",
    "    return seq_dict, score_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dr_disr_num:12264\n"
     ]
    }
   ],
   "source": [
    "# add guide mfe energy and direct repeat disruption feature\n",
    "guide_linfold_c = '../data/Linearfold_output/essential_guides_output.txt'\n",
    "#lin_seq_dict, lin_result_dict = parse_guide_linearfold_fasta_into_dict_contrafold(guide_linfold_c)\n",
    "lin_seq_dict, lin_result_dict = parse_guide_linearfold_fasta_into_dict_contrafold_rna(guide_linfold_c)\n",
    "\n",
    "linearfold_dr = [lin_seq_dict[guide][0:36] for guide in df['guide'].values]\n",
    "ref_dr = '.....(((((((.(((....))).))))))).....'\n",
    "dr_disr_num =0\n",
    "for jj in range(num_examples):\n",
    "    if linearfold_dr[jj] == ref_dr:\n",
    "        linearfold_dr[jj] = 0\n",
    "    else:\n",
    "        linearfold_dr[jj] = 1\n",
    "        dr_disr_num += 1\n",
    "print('dr_disr_num:'+str(dr_disr_num))  \n",
    "\n",
    "linearfold_vals = [lin_result_dict[guide] for guide in df['guide'].values]\n",
    "for ii in range(num_examples):\n",
    "    linearfold_vals[ii] = linearfold_vals[ii]-6.48\n",
    "    \n",
    "df['linearfold_vals_new']=linearfold_vals\n",
    "df['linearfold_dr_flag_new']= linearfold_dr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min(linearfold_vals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>linearfold_vals</th>\n",
       "      <th>linearfold_vals_new</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>linearfold_vals</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>linearfold_vals_new</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     linearfold_vals  linearfold_vals_new\n",
       "linearfold_vals                  1.0                  1.0\n",
       "linearfold_vals_new              1.0                  1.0"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['linearfold_vals','linearfold_vals_new']].corr(method='pearson')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>linearfold_vals</th>\n",
       "      <th>linearfold_vals_new</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>linearfold_vals</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>linearfold_vals_new</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     linearfold_vals  linearfold_vals_new\n",
       "linearfold_vals                  1.0                  1.0\n",
       "linearfold_vals_new              1.0                  1.0"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['linearfold_vals','linearfold_vals_new']].corr(method='spearman')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## isoform target percent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "67542\n",
      "60847\n"
     ]
    }
   ],
   "source": [
    "refseq_gene_transcript = '../data/Refseq_all/Refseq_gene_info.txt'\n",
    "df_info = pd.read_csv(refseq_gene_transcript,sep='\\t')\n",
    "#coding transcripts and genes\n",
    "df_info_coding = df_info[df_info['name'].astype(str).str.startswith('NM')]\n",
    "print(len(df_info_coding['name2'].values))\n",
    "\n",
    "#make gene_dic of transcripts, coding\n",
    "gene_transcript = {} #key:gene, value: transcript list\n",
    "transcript_list =[]\n",
    "for index, row in df_info_coding.iterrows():\n",
    "    gene_n = row['name2']\n",
    "    transcript_id = row['name']\n",
    "    transcript_list.append(transcript_id)\n",
    "    if gene_n in gene_transcript.keys():\n",
    "        if transcript_id not in gene_transcript[gene_n]:\n",
    "            gene_transcript[gene_n].append(transcript_id)\n",
    "    else:\n",
    "        gene_transcript[gene_n]=[transcript_id]\n",
    "        \n",
    "transcript_list = list(set(transcript_list))\n",
    "print(len(transcript_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get sequence for coding transcripts - not predicted\n",
    "transcript_seq = {} #key: transcript, value: sequence\n",
    "input_fasta = \"../data/Refseq_all/Refseqall_gene_seq.fa\" \n",
    "for record in SeqIO.parse(input_fasta, \"fasta\"):\n",
    "    sequence = record.seq\n",
    "    transcript_ID = record.id\n",
    "    transcript_ID = transcript_ID.split('ncbiRefSeq_')[1]\n",
    "    if (transcript_ID.startswith('NM')) and (transcript_ID not in transcript_seq.keys()):\n",
    "        transcript_seq[transcript_ID]= sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>guide</th>\n",
       "      <th>transcript_list</th>\n",
       "      <th>target_pos_list</th>\n",
       "      <th>target transcript number</th>\n",
       "      <th>target pos num</th>\n",
       "      <th>target_transcript_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGTCTCCAGACTCCACACCGGAAAGAGAG</th>\n",
       "      <td>TCGTCTCCAGACTCCACACCGGAAAGAGAG</td>\n",
       "      <td>[NM_001025071.2, NM_001025070.2, NM_005617.4]</td>\n",
       "      <td>[0, 0, 0]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTCGTCTCCAGACTCCACACCGGAAAGAGA</th>\n",
       "      <td>GTCGTCTCCAGACTCCACACCGGAAAGAGA</td>\n",
       "      <td>[NM_001025071.2, NM_001025070.2, NM_005617.4]</td>\n",
       "      <td>[1, 1, 1]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGTCGTCTCCAGACTCCACACCGGAAAGAG</th>\n",
       "      <td>CGTCGTCTCCAGACTCCACACCGGAAAGAG</td>\n",
       "      <td>[NM_001025071.2, NM_001025070.2, NM_005617.4]</td>\n",
       "      <td>[2, 2, 2]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACGTCGTCTCCAGACTCCACACCGGAAAGA</th>\n",
       "      <td>ACGTCGTCTCCAGACTCCACACCGGAAAGA</td>\n",
       "      <td>[NM_001025071.2, NM_001025070.2, NM_005617.4]</td>\n",
       "      <td>[3, 3, 3]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACGTCGTCTCCAGACTCCACACCGGAAAG</th>\n",
       "      <td>CACGTCGTCTCCAGACTCCACACCGGAAAG</td>\n",
       "      <td>[NM_001025071.2, NM_001025070.2, NM_005617.4]</td>\n",
       "      <td>[4, 4, 4]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         guide  \\\n",
       "TCGTCTCCAGACTCCACACCGGAAAGAGAG  TCGTCTCCAGACTCCACACCGGAAAGAGAG   \n",
       "GTCGTCTCCAGACTCCACACCGGAAAGAGA  GTCGTCTCCAGACTCCACACCGGAAAGAGA   \n",
       "CGTCGTCTCCAGACTCCACACCGGAAAGAG  CGTCGTCTCCAGACTCCACACCGGAAAGAG   \n",
       "ACGTCGTCTCCAGACTCCACACCGGAAAGA  ACGTCGTCTCCAGACTCCACACCGGAAAGA   \n",
       "CACGTCGTCTCCAGACTCCACACCGGAAAG  CACGTCGTCTCCAGACTCCACACCGGAAAG   \n",
       "\n",
       "                                                              transcript_list  \\\n",
       "TCGTCTCCAGACTCCACACCGGAAAGAGAG  [NM_001025071.2, NM_001025070.2, NM_005617.4]   \n",
       "GTCGTCTCCAGACTCCACACCGGAAAGAGA  [NM_001025071.2, NM_001025070.2, NM_005617.4]   \n",
       "CGTCGTCTCCAGACTCCACACCGGAAAGAG  [NM_001025071.2, NM_001025070.2, NM_005617.4]   \n",
       "ACGTCGTCTCCAGACTCCACACCGGAAAGA  [NM_001025071.2, NM_001025070.2, NM_005617.4]   \n",
       "CACGTCGTCTCCAGACTCCACACCGGAAAG  [NM_001025071.2, NM_001025070.2, NM_005617.4]   \n",
       "\n",
       "                               target_pos_list  target transcript number  \\\n",
       "TCGTCTCCAGACTCCACACCGGAAAGAGAG       [0, 0, 0]                         3   \n",
       "GTCGTCTCCAGACTCCACACCGGAAAGAGA       [1, 1, 1]                         3   \n",
       "CGTCGTCTCCAGACTCCACACCGGAAAGAG       [2, 2, 2]                         3   \n",
       "ACGTCGTCTCCAGACTCCACACCGGAAAGA       [3, 3, 3]                         3   \n",
       "CACGTCGTCTCCAGACTCCACACCGGAAAG       [4, 4, 4]                         3   \n",
       "\n",
       "                                target pos num  target_transcript_percent  \n",
       "TCGTCTCCAGACTCCACACCGGAAAGAGAG               1                        1.0  \n",
       "GTCGTCTCCAGACTCCACACCGGAAAGAGA               1                        1.0  \n",
       "CGTCGTCTCCAGACTCCACACCGGAAAGAG               1                        1.0  \n",
       "ACGTCGTCTCCAGACTCCACACCGGAAAGA               1                        1.0  \n",
       "CACGTCGTCTCCAGACTCCACACCGGAAAG               1                        1.0  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make guide dic for essential genes in our screen\n",
    "gene_ess = ['RPS14', 'CDC5L', 'POLR2I', 'RPS7', 'XAB2', 'RPS19BP1', 'RPL23A', 'SUPT6H', 'PRPF31', 'U2AF1', 'PSMD7',\n",
    "         'HSPE1', 'RPS13', 'PHB', 'RPS9', 'EIF5B', 'RPS6', 'RPS11', 'SUPT5H', 'SNRPD2', 'RPL37', 'RPSA', 'COPS6',\n",
    "         'DDX51', 'EIF4A3', 'KARS1', 'RPL5', 'RPL32', 'SF3A1', 'RPS3A', 'SF3B3', 'POLR2D', 'RPS15A', 'RPL31', 'PRPF19',\n",
    "         'SF3B2', 'RPS4X', 'CSE1L', 'RPL6', 'COPZ1', 'PSMB2', 'RPL7', 'PHB2', 'ARCN1', 'RPA2', 'NUP98', 'RPS3', 'EEF2',\n",
    "         'USP39', 'PSMD1', 'NUP93', 'AQR', 'RPL34', 'PSMA1', 'RPS27A']\n",
    "print(len(gene_ess))\n",
    "guide_dic ={}\n",
    "\n",
    "for gene in gene_ess:\n",
    "    gene_guide_dic = {} #gene-specific guide dic\n",
    "    trans_list_all = gene_transcript[gene] #may include non-coding transcripts\n",
    "    trans_list = [] #only include coding transcripts that start with 'NM'\n",
    "    for tt in trans_list_all:\n",
    "        if tt.startswith('NM'):\n",
    "            trans_list.append(tt)\n",
    "    trans_num = len(trans_list) #isoform total number\n",
    "    for trans_id in trans_list:\n",
    "        if trans_id not in transcript_seq.keys():\n",
    "            print('version different')\n",
    "            id_short = trans_id.split('.')[0] #use one without version id\n",
    "            id_versions = 0 # how many isoform with different versions for this short id\n",
    "            for t in transcript_seq.keys():\n",
    "                if id_short in t: #transcript id matches, version different\n",
    "                    id_versions += 1\n",
    "                    #trans_id = t\n",
    "                    #design the guides\n",
    "                    trans_seq = transcript_seq[t]\n",
    "                    for i in range(0,len(trans_seq)-29):\n",
    "                        target = trans_seq[i:i+30]\n",
    "                        spacer = target.reverse_complement()\n",
    "                        if str(spacer) not in gene_guide_dic.keys(): #first occur\n",
    "                            gene_guide_dic[str(spacer)]=[str(spacer),[t],[i],1,1] \n",
    "                                                    #spacer, transcript_list,pos_list,target_trans_num, target_pos_num_for_one_transcript\n",
    "                        else:\n",
    "                            if t not in gene_guide_dic[str(spacer)][1]:\n",
    "                                gene_guide_dic[str(spacer)][1].append(t)\n",
    "                                gene_guide_dic[str(spacer)][2].append(i)\n",
    "                                if id_versions == 1: #only add count if it's the first isoform for this transcript\n",
    "                                    gene_guide_dic[str(spacer)][3] += 1 #target transcript number\n",
    "                            else: #guide targeting the same transcript\n",
    "                                gene_guide_dic[str(spacer)][4] += 1\n",
    "                                \n",
    "            \n",
    "        else:\n",
    "            #design the guides\n",
    "            trans_seq = transcript_seq[trans_id]\n",
    "            for i in range(0,len(trans_seq)-29):\n",
    "                target = trans_seq[i:i+30]\n",
    "                spacer = target.reverse_complement()\n",
    "                if str(spacer) not in gene_guide_dic.keys(): #first occur\n",
    "                    gene_guide_dic[str(spacer)]=[str(spacer),[trans_id],[i],1,1]\n",
    "                                        #spacer, transcript_list,pos_list,target_trans_num, target_pos_num_for_one_transcript\n",
    "                else:\n",
    "                    if trans_id not in gene_guide_dic[str(spacer)][1]:\n",
    "                        gene_guide_dic[str(spacer)][1].append(trans_id)\n",
    "                        gene_guide_dic[str(spacer)][2].append(i)\n",
    "                        gene_guide_dic[str(spacer)][3] += 1 #target transcript number\n",
    "                    else: #guide targeting the same transcript\n",
    "                        gene_guide_dic[str(spacer)][4] += 1\n",
    "                        \n",
    "    for guide in gene_guide_dic.keys():\n",
    "        if guide not in guide_dic.keys():\n",
    "            target_percent = float(gene_guide_dic[guide][3])/trans_num\n",
    "            guide_dic[guide]= gene_guide_dic[guide]+[target_percent]\n",
    "        else: #guide target other genes\n",
    "            guide_dic.pop(guide) #remove the guide with off targets\n",
    "            \n",
    "\n",
    "guide_iso = pd.DataFrame.from_dict(guide_dic, orient='index',columns=['guide','transcript_list','target_pos_list','target transcript number','target pos num','target_transcript_percent'])\n",
    "guide_iso.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('essential_guide_target_percent.csv','w') as outf:\n",
    "    writer = csv.writer(outf)\n",
    "    writer.writerow(['guide','transcript_list','target_pos_list','target transcript number','target pos num','target_transcript_percent'])\n",
    "    for g in guide_dic.keys():\n",
    "        writer.writerow(guide_dic[g])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## position flags and floats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create dictionary of transcripts\n",
    "transcript_dictionary = {}\n",
    "gene_len = 'lengths.csv'\n",
    "gene_cds = \"CDS.csv\"\n",
    "\n",
    "#CDS position\n",
    "with open(gene_cds, 'r') as f:\n",
    "\tfor row in csv.reader(f.read().splitlines()):\n",
    "\t\ttranscript = str(row[1])\n",
    "\t\ttranscript_dictionary[transcript] = row[2:4] #CDS position\n",
    "        \n",
    "#gene length\n",
    "with open(gene_len, 'r') as f:\n",
    "\tfor row in csv.reader(f.read().splitlines()):\n",
    "\t\ttranscript = str(row[1])\n",
    "\t\ttranscript_dictionary[transcript] = transcript_dictionary[transcript]+[row[-1]] #gene length\n",
    "        \n",
    "print(len(transcript_dictionary))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#CDS, 5'UTR, 3'UTR pos floats\n",
    "output_dictionary = {}\n",
    "#is_CDS = []\n",
    "#is_5UTR = []\n",
    "#is_3UTR = []\n",
    "CDS_position = []\n",
    "UTR5_position = []\n",
    "UTR3_position = []\n",
    "\n",
    "\n",
    "#CDS, 5'UTR, 3'UTR\n",
    "for guide, attributes in guide_dictionary.items():\n",
    "\ttranscript = attributes[2] \n",
    "\tposition = int(attributes[-20])\n",
    "#\tprint(position)\n",
    "#\tbreak\n",
    "\tCDS_start = int(transcript_dictionary[transcript][0])\n",
    "\tCDS_end = int(transcript_dictionary[transcript][1])\n",
    "\tgene_length = int(transcript_dictionary[transcript][2]) \n",
    "\tif position > CDS_start: #if the guide starts after the start of the CDS\n",
    "\t\tif position < CDS_end : # and before the end of the CDS\n",
    "#\t\t\tis_CDS.append(guide_dictionary[guide] + [1])\n",
    "\t\t\trel_position = float(position - CDS_start) / float(CDS_end - CDS_start)\n",
    "\t\t\toutput_dictionary[guide]= guide_dictionary[guide] + [0,rel_position,0]\n",
    "\t\t\tCDS_position.append(guide_dictionary[guide] + [rel_position])\n",
    "#\t\t\tis_3UTR.append(guide_dictionary[guide] + [0])\n",
    "#\t\t\tis_5UTR.append(guide_dictionary[guide] + [0])\n",
    "\t\t\tUTR3_position.append(guide_dictionary[guide] + [0])\n",
    "\t\t\tUTR5_position.append(guide_dictionary[guide] + [0])\n",
    "\t\telse:\n",
    "#\t\t\tis_3UTR.append(guide_dictionary[guide] + [1])\n",
    "#\t\t\tis_5UTR.append(guide_dictionary[guide] + [0])\n",
    "#\t\t\tis_CDS.append(guide_dictionary[guide] + [0])\n",
    "\t\t\trel_position = float(position - CDS_end) / float(gene_length-CDS_end)\n",
    "\t\t\toutput_dictionary[guide]= guide_dictionary[guide] + [0,0,rel_position]\n",
    "\t\t\tUTR3_position.append(guide_dictionary[guide] + [rel_position]) \n",
    "\t\t\tCDS_position.append(guide_dictionary[guide] + [0])\n",
    "\t\t\tUTR5_position.append(guide_dictionary[guide] + [0])\n",
    "\telse: \n",
    "#\t\tis_5UTR.append(guide_dictionary[guide] + [1])\n",
    "#\t\tis_CDS.append(guide_dictionary[guide] + [0])\n",
    "#\t\tis_3UTR.append(guide_dictionary[guide] + [0])\n",
    "\t\tCDS_position.append(guide_dictionary[guide] + [0])\n",
    "\t\tUTR3_position.append(guide_dictionary[guide] + [0])\n",
    "\t\trel_position = float(position)/float(CDS_start)\n",
    "\t\tUTR5_position.append(guide_dictionary[guide] + [rel_position])\n",
    "\t\toutput_dictionary[guide]= guide_dictionary[guide] + [rel_position,0,0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## flanking sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "guides_pos = {}\n",
    "\n",
    "input_fasta = \"essential_transcripts.fasta\"\n",
    "\n",
    "for record in SeqIO.parse(input_fasta, \"fasta\"):\n",
    "\tsequence = record.seq\n",
    "\ttranscript_ID = record.id\n",
    "\ttitle = str(record.description)\n",
    "\n",
    "\t#find the gene symbol\n",
    "\ta = title.find(\"(\")\n",
    "\tb = title.find(\")\")\n",
    "#\tgene_symbol = title.split( )[1]\n",
    "\tgene_symbol = title[a+1:b]\n",
    "\n",
    "\t#design the guides\n",
    "\tfor i in range(0,len(record.seq)-29):\n",
    "\t\ttarget = sequence[i:i+30]\n",
    "\t\tspacer = target.reverse_complement()\n",
    "\t\tif (i != 0) and (i != len(record.seq)-30):\n",
    "\t\t\tleft_seq_5 = sequence[max(0,(i-5)):i]\n",
    "\t\t\tright_seq_5 = sequence[i+30:min((i+35),len(record.seq))]\n",
    "\t\t\tleft_seq_10 = sequence[max(0,(i-10)):i]\n",
    "\t\t\tright_seq_10 = sequence[i+30:min((i+40),len(record.seq))]\n",
    "\t\t\tleft_seq_15 = sequence[max(0,(i-15)):i]\n",
    "\t\t\tright_seq_15 = sequence[i+30:min((i+45),len(record.seq))]\n",
    "\t\t\tleft_seq_20 = sequence[max(0,(i-20)):i]\n",
    "\t\t\tright_seq_20 = sequence[i+30:min((i+50),len(record.seq))]\n",
    "\t\t\tleft_seq_100 = sequence[max(0,(i-100)):i]\n",
    "\t\t\tright_seq_100 = sequence[i+30:min((i+130),len(record.seq))]\n",
    "\t\telif i == 0:\n",
    "\t\t\tleft_seq_5 = ''\n",
    "\t\t\tright_seq_5 = sequence[i+30:min((i+35),len(record.seq))]\n",
    "\t\t\tleft_seq_10 = ''\n",
    "\t\t\tright_seq_10 = sequence[i+30:min((i+40),len(record.seq))]\n",
    "\t\t\tleft_seq_15 = ''\n",
    "\t\t\tright_seq_15 = sequence[i+30:min((i+45),len(record.seq))]\n",
    "\t\t\tleft_seq_20 = ''\n",
    "\t\t\tright_seq_20 = sequence[i+30:min((i+50),len(record.seq))]\n",
    "\t\t\tleft_seq_100 = ''\n",
    "\t\t\tright_seq_100 = sequence[i+30:min((i+130),len(record.seq))]\n",
    "\t\telif i == len(record.seq)-30:  \n",
    "\t\t\tleft_seq_5 = sequence[max(0,(i-5)):i]\n",
    "\t\t\tright_seq_5 = ''\n",
    "\t\t\tleft_seq_10 = sequence[max(0,(i-10)):i]\n",
    "\t\t\tright_seq_10 = ''\n",
    "\t\t\tleft_seq_15 = sequence[max(0,(i-15)):i]\n",
    "\t\t\tright_seq_15 = ''\n",
    "\t\t\tleft_seq_20 = sequence[max(0,(i-20)):i]\n",
    "\t\t\tright_seq_20 = ''\n",
    "\t\t\tleft_seq_100 = sequence[max(0,(i-100)):i]\n",
    "\t\t\tright_seq_100 = ''\n",
    "\t\tnearby_seq_all_5 = left_seq_5 + target + right_seq_5 #target seq+flanks\n",
    "\t\tnearby_seq_all_10 = left_seq_10 + target + right_seq_10\n",
    "\t\tnearby_seq_all_15 = left_seq_15 + target + right_seq_15 \n",
    "\t\tnearby_seq_all_20 = left_seq_20 + target + right_seq_20 \n",
    "\t\tnearby_seq_all_100 = left_seq_100 + target + right_seq_100 #target seq+flanks\n",
    "#\t\tnearby_seq_all_rv  = nearby_seq_all.reverse_complement()\n",
    "#\t\tguide_loc_flag = np.zeros(len(nearby_seq_all))\n",
    "#\t\tguide_loc = nearby_seq_all_rv.find(spacer)\n",
    "#\t\tguide_loc_flag[guide_loc:guide_loc+30]=1 # 1 means where the guide is in this flag vector\n",
    "\t\tif str(spacer) not in guides_pos:\n",
    "#\t\t\tguides_pos[spacer]= [gene_symbol,transcript_ID,i,str(left_seq),str(right_seq),str(nearby_seq_all),str(nearby_seq_all_rv),guide_loc_flag]\n",
    "\t\t\tguides_pos[spacer]= [gene_symbol,transcript_ID,i,str(target),str(left_seq_5),str(right_seq_5),str(nearby_seq_all_5),str(nearby_seq_all_10),str(nearby_seq_all_15),str(nearby_seq_all_20),str(nearby_seq_all_100)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#target with flank 15 nt\n",
    "#native\n",
    "nearby_seq_all = df['nearby_seq_all_15'].values\n",
    "with open(\"Linearfold_input/essential_target_flank15_input.fasta\", 'w') as file:\n",
    "    for i in range(num_examples):\n",
    "        file.write(\">guide #\" + str(i + 1) + \"\\n\")\n",
    "        file.write(nearby_seq_all[i] + \"\\n\")\n",
    "        \n",
    "# with constraints\n",
    "guide_rv = df['target_seq'].values\n",
    "with open(\"Linearfold_input/essential_target_constraints_flank15_input.fasta\", 'w') as file:\n",
    "    for i in range(num_examples):\n",
    "        file.write(\">guide #\" + str(i + 1) + \"\\n\")\n",
    "        file.write(nearby_seq_all[i] + \"\\n\")\n",
    "        target_index = nearby_seq_all[i].find(guide_rv[i])\n",
    "        constraints = \"?\"*target_index + \".\"*30 + \"?\"*(len(nearby_seq_all[i])-30-target_index)\n",
    "        file.write(constraints + \"\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
